### ISSUE TITLE
Inconsitent generics infererence between ECJ and the JDK compiler

### ISSUE THREAD

**Issue by alienisty** [eclipse-jdt/eclipse.jdt.core#366]
There is a case where the JDK successfully infers the generics where the eclipse compiler emits an error.
I'm using Adoptium JDK 11.0.13+8 and eclipse 4.22.
The following file is the simplest reproduction of the problem I was able to muster:
```java
public class A {
  Collection<S<? extends X>> values() { return List.of(); }
  void handle() { problemsWith(values()); } // eclipse complains here
  void problemsWith(Collection<S<?>> values) {}
}

class S<T extends Y> {}

class X {}

class Y extends X {}
```
I've also tried to compile the same with Adoptium JDK 17.0.2+8 and the class still compiled fine.
I'm not sure on which side is the bug though.

**Comment by stephan-herrmann** [comment]
Yeah, we never settled with Oracle over the exact inference rules when wild cards are involved. They acknowledged that javac does more than what's written in JLS, but we never succeeded to agree on what a compiler really *should* do. My guess is that first JLS needs to be improved before we could possibly make progress.

Previous discussions:
* https://bugs.eclipse.org/bugs/show_bug.cgi?id=448793 (with lots of other bugs linked)
* https://bugs.openjdk.org/browse/JDK-8054721 (opened 2014, stalled since 2017)

**Comment by trancexpress** [comment]
See also #837.
